libs/core/thread_support/include/hpx/thread_support/spinlock.hpp

```diff
@@ -45,15 +45,18 @@ namespace hpx { namespace util { namespace detail {
         void lock() noexcept
         {
             // Optimistically assume the lock is free on the first try
-            if (!m.exchange(true, std::memory_order_acquire))
+            if (!m.load(std::memory_order_relaxed) &&
+                !m.exchange(true, std::memory_order_acquire))
                 return;
 
             // Wait for lock to be released without generating cache misses
+            // Similar implementation to hpx::lcos::local::spinlock
             unsigned k = 0;
-            do
+            while (m.load(std::memory_order_relaxed) ||
+                m.exchange(true, std::memory_order_acquire))
             {
                 yield_k(k++);
-            } while (!try_lock());
+            }
```

Comment:
A `do{} while()` would be more appropriate as we know that the lock can't be acquired if `k == 0`. Also, wy not call `try_lock()` instead?

Reply:
`while()` vs `do{} while()` : `yield_k` for `k == 0` is a no-op so I felt both ways of representing are equivalent. This can be changed however if you'd like it to be a `do{} while()` without any change in performance.

why not call `try_lock`? The nature of the return statement means that `!m.load(std::memory_order_relaxed)` will always compute to false in cases where the lock has been previously acquired. This means that `!try_lock()` will essentially trigger an exchange statement every iteration. This can be mitigated by doing something like:

```
while (true)
{
    if (try_lock())
        return;
    yield_k(k++);
}
```

Would you want me to change the code to look like above?

Reply:
It could also be turned into a `do{} while()` using which we can get rid of the above optimistic try_lock. I think I'll convert it to `do{} while()` for better code clarity.